### DISCUSSION TITLE
[Program] Calculating slope of ground at a given point, and direction/orientation of slope

### DISCUSSION THREAD

**Discussion by cbot-coder** [colobot/colobot#1369] [Workshop]
I'm trying to program a pathfinding algorithm for Saari 1 (the Uranium mountain mission). One element I needed was a `slope()` function to determine the slope of the ground at a given point. There are many ways to do that, one of which obviously is to take two points and calculate the `atan` of the difference in `topo()` for them, over the horizontal distance between them. But selecting the two points so that they show the maximum slope at that point requires sampling several points at least. Then, I started trying to find other ways to do it. Poking around the C++ source code, I found an approach that gives you the exact slope at a given point using some 3D plane mathematics. I tweaked it and simplified it so that it executes as fast as possible in CBOT code. You give it a point, and it picks two other close points from which to sample the `topo()` values, such that the math simplifies many terms so they drop out, making the calculations a lot simpler than the general case using any 3 given points.

In addition, because it has all the information to return you the direction/orientation of the slope (the direction you would slide if standing on the slope), I wanted it to return that at the same time, thus the `float[]` return type. Then, since we already grabbed the `topo()` at the given point, we go ahead and return that as well. The return value is explained well in the comments below:

```c++
  // Returns the slope, the direction (orientation) of the slope, and
  // the topo (z-height of the ground) at the given point. The return
  // value is a float[] array with the following values:
  // - [0] is the slope (e.g. 0 = flat, 15 = slope of 15 degrees, etc.)
  // - [1] is the orientation of downhill (the direction you would slide)
  // - [2] is the topo (z-height), just in case it is needed
  //
  // This is a mathematically simplified and thus much faster version
  // of the full math that would be required to dermine the slope from
  // any given three points. By choosing the other two points in a
  // certain way, the math simplifies down quite a bit, eliminating or
  // simplifying many terms because they are zero or directly related
  // to the 'step' parameter, etc.  If you need to figure out how it
  // was derived, look at CTerrain::GetNormal, CTerrain::GetFineSlope
  // in terrain.cpp, Vector::CrossProduct in vector.h in the source
  // code for Colobot Gold Edition, and for more background on the
  // math, look at https://www.youtube.com/watch?v=rL9UXzZYYo4
  //
  // The 'step' parameter allows callers to adjust the size of the
  // triangle used to calculate slope. A larger value gives a coarser
  // slope that is probably the average of a larger area. A smaller
  // value means to measure the slope at a more precise location
  // around the given point. If you want to sample the slope of a
  // larger area around a given point, it may be good to call twice:
  // once with a positive step and once with a negative step (so the
  // two triangles used to calculate are on opposite sides of the
  // given point), then average the slopes.
  public float[] slopeDirTopo(point p, float step = 0.1) {
    float pz = topo(p),
          p2x = step * (pz - topo(p.x + step, p.y)),
          p2y = step * (pz - topo(p.x, p.y + step)),
          retval[];
    retval[0] = atan2(sqrt(p2x * p2x + p2y * p2y), step * step); // slope
    retval[1] = atan2(p2y, p2x); // orientation of slope
    retval[2] = pz;
    return retval;
  }
```

If you can think of any improvements, or other ways to use this, or have any other feedback or ideas, please comment below. Thanks!

**Comment by Zemogiter** [comment]
Would that work?
```
if (slopeDirTopo(position, 0.1) > some_value && this.category == WheeledGrabber) 
{
    message("Terrain too steep, finding alternative route");
    //whatever code used to look for alternative route
}
```
